### PR TITLE
test/keepalive: Improve TestKeepaliveClientClosesWithActiveStreams 

### LINKED ISSUE #6099: Fix possible race in Keepalive test: KeepaliveClientClosesWithActiveStreams
KeepaliveClientClosesWithActiveStreams test creates a server which does not respond to keepalive pings, and makes sure that the client closes the transport even when there is an active stream. However there is a possibility of a race here. 

In the test we assign a [large enough](https://github.com/grpc/grpc-go/blob/master/internal/transport/keepalive_test.go#L329) `ClientParameters.Time`.  However this does not guarantee that a create can be created successfully before the timeout hits. This is because with the `NoPingServer` we do not have a configurable way to switch between a ping/noping server. 

This way we can verify that the "no ping server" actually respond to pings until you set a flag, at which point it stops responding and the timeout kicks in to close the transport.

### PR DESCRIPTION
**Pull request by tobotg**
Setup a ping/noping switchable server to use in KeepaliveClientClosesWithActiveStreams test

Fixes #6099

RELEASE NOTES: N/A

### CODE REVIEW

**Comment by tobotg** [comment]
@arvindbr8 Hello, I'm unsure about what caused **[Testing / vet-proto (pull_request)](https://github.com/grpc/grpc-go/actions/runs/4984005112/jobs/8921766996?pr=6282)** above to fail and how to fix it for this PR. Is it something you could help me with? Thanks in advance.

**Comment by dfawley** [comment]
> Hello, I'm unsure about what caused Testing / vet-proto (pull_request) above to fail

This one is not actionable in your PR.  And I believe `extras` should be fixed if you fetch master@HEAD and rebase+force-push (or merge into?) your branch.

**Comment by tobotg** [comment]
Thanks @dfawley It helped.
However, are Test/DialWaitsForServerSettingsAndFails and Test/FullHandshake known flaky? They failed in the last run (one in [Testing / tests (tests, 1.20) (pull_request)](https://github.com/grpc/grpc-go/actions/runs/4988867731/jobs/8932073191?pr=6282) and the other in [Testing / tests (tests, 1.20, arm64) (pull_request)](https://github.com/grpc/grpc-go/actions/runs/4988867731/jobs/8932073496?pr=6282) ), but they are not related to the fix I pushed. Any thoughts?

File: internal/transport/transport_test.go
Comment (arvindbr8) [inline internal/transport/transport_test.go:1]: I would prefer parameterizing the existing `NoPingServer` setup using `setUpWithNoPingServer` to be configurable. Reimplementing `HandleStreams` doesnt seem right to me. Also I'm confused why we had to update `handleStreamInvalidHeaderField` for handling the keepalive tests. 

We should try to extend the `setUpWithNoPingServer`. Also take a look at [this comment on a closed PR ](https://github.com/grpc/grpc-go/pull/6168#issuecomment-1495388746) for some additional context.

**Review by arvindbr8 — CHANGES_REQUESTED** [review]

**Comment by arvindbr8** [comment]
Hey @tobotg 
Test/DialWaitsForServerSettingsAndFails and Test/FullHandshake are not known flakes. I've triggered a re-run. lets see how that goes.

**Comment by tobotg** [comment]
Hey @arvindbr8 all the tests passed now, so it's.
btw, thanks for the review. I'll look into it and get back here.

**Comment by tobotg** [comment]
I've been very busy lately, hopefully I'll be able to get back to this during the weekend.

File: internal/transport/transport_test.go
Comment (tobotg) [reply internal/transport/transport_test.go:1]: > I would prefer parameterizing the existing `NoPingServer` setup using `setUpWithNoPingServer` to be configurable. Reimplementing `HandleStreams` doesnt seem right to me. Also I'm confused why we had to update `handleStreamInvalidHeaderField` for handling the keepalive tests.
> 
> We should try to extend the `setUpWithNoPingServer`. Also take a look at [this comment on a closed PR ](https://github.com/grpc/grpc-go/pull/6168#issuecomment-1495388746) for some additional context.

Got some time this morning to look into this, and I'm going to submit changes. I hope I got your point. Let me know when you get a chance. Thanks

**Comment by tobotg** [comment]
Hey @arvindbr8 , I gave it another try. Please have a look when you get a chance. Thanks

**Comment by dfawley** [comment]
@arvindbr8 is out of town for a couple weeks; will resume when he returns.

**Comment by arvindbr8** [comment]
Hey @tobotg, thanks for the revision. I will take a look at it shortly.

File: internal/transport/keepalive_test.go
Comment (arvindbr8) [inline internal/transport/keepalive_test.go:318]: When we did not have a way to setup a configurable server (like you have added in this diff), we set a "large enough"  time + timeout for KeepaliveParams (`500ms`) to ensure we could atleast create one successful stream before the keepalive timeout kicks and client closes the connection. 

Now since we have more control, i believe we should be able to reduce the timeouts and keep it as low as possible. 

How about setting Time: 10ms and Timeout: 10ms.

File: internal/transport/keepalive_test.go
Comment (arvindbr8) [inline internal/transport/keepalive_test.go:344]: I dont think we need this now. It should be good enough to verify that we have atleast one active stream and when the server stops responding to keepalive, stream creation starts erroring.

File: internal/transport/transport_test.go
Comment (arvindbr8) [inline internal/transport/transport_test.go:481]: Why do we need this check now?

File: internal/transport/transport_test.go
Comment (arvindbr8) [inline internal/transport/transport_test.go:487]: is `framer.ErrorDetail()` inclusive of `err` ?

File: internal/transport/transport_test.go
Comment (arvindbr8) [inline internal/transport/transport_test.go:512]:
```suggestion
					// ack the ping if in responsive mode.
```

File: internal/transport/transport_test.go
Comment (arvindbr8) [inline internal/transport/transport_test.go:497]: nit: How about?
```suggestion
				respondToPing := false
```

File: internal/transport/transport_test.go
Comment (arvindbr8) [inline internal/transport/transport_test.go:467]:
```suggestion
func setUpWithNoPingServerWithOptions(t *testing.T, copts ConnectOptions, connCh chan net.Conn, respModeChan chan bool) (*http2Client, func()) {
```

Comment (arvindbr8) [inline internal/transport/transport_test.go:467]: Please add a comment to the helper. This would be help make more sense of what `WithOptions` means her.

**Review by arvindbr8 — CHANGES_REQUESTED** [review]

**Comment by arvindbr8** [comment]
thanks @tobotg, Changes looks better now. A few minor comments and nits. Please take a look and let us know.

File: internal/transport/transport_test.go
Comment (tobotg) [reply internal/transport/transport_test.go:486]: Yes, it is and it's more detailed

File: internal/transport/transport_test.go
Comment (tobotg) [reply internal/transport/transport_test.go]: Done

**Comment by tobotg** [comment]
Thanks @arvindbr8 for the review.
I had a look and addressed the changes you requested. Please let me know how it looks now.

**Review by arvindbr8 — CHANGES_REQUESTED** [review]
I had one more comment about how the respModeChan is being read and one minor update to a comment.

File: internal/transport/transport_test.go
Comment (arvindbr8) [reply internal/transport/transport_test.go:508]: On another look, I think we can do without the `ok` check. Also since we have a non blocking read, do we want the channel to have a buffer?

Comment (tobotg) [reply internal/transport/transport_test.go:508]: Reading on the empty channel leads to `default` in the `select` clause. The `ok` here is used to know when the channel is closed and all sent data has been received (empty **and** closed), and then to exit the goroutine (to avoid an infinite loop and a leaking goroutine). I guess `framer.ReadFrame()` could lead to exit, but I can't just leave this code keep reading a closed channel. Would you suggest a different approach?

Comment (arvindbr8) [reply internal/transport/transport_test.go:508]: I agree and it makes sense. I'm okay with the correct approach.

**Comment by arvindbr8** [comment]
@tobotg, Please let us know when you want us to take another look at this.

**Comment by tobotg** [comment]
@arvindbr8 Thanks for the review (and I added your comment). Please take another look when you get a chance.

File: internal/transport/transport_test.go
Comment (arvindbr8) [inline internal/transport/transport_test.go:511]: minor nit
```suggestion
					case mdoe, ok := <-respModeCh:
						if !ok {
							return
						}
						respondToPing = mode
```

**Review by arvindbr8 — APPROVED** [review]
LGTM. One minor naming nit. cc: @dfawley, could you please take a look for a second pair of eyes?

File: internal/transport/transport_test.go
Comment (tobotg) [reply internal/transport/transport_test.go]: Thanks @arvindbr8 . I didn't use your suggestion directly as there was a typo, but I just made the change you suggested (replace `data` by `mode`).

Comment (arvindbr8) [reply internal/transport/transport_test.go]: thanks!

File: internal/transport/transport_test.go
Comment (tobotg) [reply internal/transport/transport_test.go]: @arvindbr8 This change seems to be failing  [tests (vet, 1.20) (pull_request)](https://github.com/grpc/grpc-go/actions/runs/5405119271/jobs/9820215415?pr=6282). Do you have any clue?

File: internal/transport/transport_test.go
Comment (arvindbr8) [inline internal/transport/transport_test.go:471]: vets are failing because of the trailing whitespace. please take a look @tobotg 
```suggestion
// * If `true` is written to respModeCh, the server will respond to ping acks. (default)
```

**Comment by tobotg** [comment]
@arvindbr8 All passed now.
@dfawley

File: internal/transport/keepalive_test.go
Comment (dfawley) [inline internal/transport/keepalive_test.go:323]: I think it would be a small amount more work, but much nicer to make `setUpWithNoPingServer` return a `server` type that has a method that allows you to change the mode.  And that can take the place of the `cancel` return parameter with a `server.stop()` method instead.

File: internal/transport/transport_test.go
Comment (dfawley) [inline internal/transport/transport_test.go:498]: A goroutine in a goroutine?

Why not push to `connCh` then do this logic?

File: internal/transport/transport_test.go
Comment (dfawley) [inline internal/transport/transport_test.go:500]: This lock doesn't seem to do anything at all

**Review by dfawley — CHANGES_REQUESTED** [review]

File: internal/transport/keepalive_test.go
Comment (tobotg) [reply internal/transport/keepalive_test.go:323]: I didn't want to change the original pattern of these utility functions, but sure thing, I can look into that when I get a chance.

File: internal/transport/transport_test.go
Comment (tobotg) [reply internal/transport/transport_test.go:498]: @dfawley Could you please clarify your comment here and what your suggestion is? I'm not following what the issue is to use a goroutine in a goroutine, I'll appreciate your help on this part. Thanks

File: internal/transport/transport_test.go
Comment (tobotg) [reply internal/transport/transport_test.go:500]: This lock protects the https://github.com/grpc/grpc-go/blob/fc7470388ccd16b300866941e8532797c329dd37/internal/transport/transport_test.go#L518

Comment (tobotg) [reply internal/transport/transport_test.go:500]: Good catch! I'll fix that. Thanks

**Comment by tobotg** [comment]
Back to it. I'll submit a new commit this week. Thanks

File: internal/transport/keepalive_test.go
Comment (tobotg) [reply internal/transport/keepalive_test.go:323]: I had a chance to look into this suggestion, and made some changes. I'll submit it later for your review. Thanks

**Comment by tobotg** [comment]
@dfawley Hello! I just submitted a new commit now, I hope it captured the feedback you shared recently. However, I'm unsure what the vet issue is here, any clue will be appreciated.Thanks

**Comment by easwars** [comment]
If you rebase your branch to master, the vet issue should get fixed. Either way, it does not affect the ability to merge the changes once approved.

**Comment by dfawley** [comment]
> If you rebase your branch to master

In general, it's best to not rebase a PR from master until the reviews are done, as it messes up github diffs and review comment threads.

In this particular case, it's not necessary at all, because the vet-proto checker is optional and the failure is because external/unrelated things are out of sync that we will fix separately.

**Comment by tobotg** [comment]
Hello @dfawley and @arvindbr8 Just in case I missed it, I wanted to check in if there is any clarification or task needed from me on this PR before review. Thanks

File: internal/transport/transport_test.go
Comment (arvindbr8) [inline internal/transport/transport_test.go:502]: I think a comment here about to callout the case if `s.respModeCh == nil` then server does not responToPing

File: internal/transport/transport_test.go
Comment (arvindbr8) [inline internal/transport/transport_test.go:566]: nit: When possible, aim for comments that will read well on an 80-column wide terminal

```suggestion
// setPingMode toggles the server between responding to ping acks and not
// responding based on the writes to the respModeCh channel.
//
// * (default) If `true` is written to respModeCh, the server will respond to
// ping acks. 
// * If `false` is written to respModeCh, the server will not respond to ping
// acks.
```

go/go-style/decisions#comment-line-length

File: internal/transport/transport_test.go
Comment (arvindbr8) [inline internal/transport/transport_test.go:584]: Lets call this `setUpWithPingModeServer` since thats the name of the server we are setting up. umm also.. im thinking we can nix `WithOptions`, since the only option we pass in here is the `switchable` bool. 

I would prefer if pass in the initial response mode to the setUp function -- we might have to pass this in to `sever.start()`. the advantage(s) i see here is that, we can initialize pingModeServer always with a respModeCh ch.  Which would also remove the conditional nil check on the respModeCh in server.start().

That would also just result in nixing `setUpWithNoPingServer` and replacing the callsites with `setUpWithPingModeServer(t, copts, false)` -- indicating no response to pings -- unless you explicitly call setPingMode().

You can also remove the conditional in setPingMode method.

File: internal/transport/transport_test.go
Comment (arvindbr8) [inline internal/transport/transport_test.go:529]: i think we can probably nix this method. The current usage can be replaced by just calling `lis.Addr().String()` when creating `NewClientTransport`

File: internal/transport/transport_test.go
Comment (arvindbr8) [inline internal/transport/transport_test.go:465]:
```suggestion
// pingModeServer is a custom server which can switch from responding to pings
// to ignoring them and vice versa using signals sent on a boolean chan
// respModeCh. If respModeCh is nil, the server would work as a no ping server.
```

Comment (arvindbr8) [reply internal/transport/transport_test.go:465]: If you follow my suggestion below, then the usage of a boolean chan to control the PingMode can be opaque to the user. Instead mention how they can just use `setPingMode` to alter the PingMode of the server.

File: internal/transport/transport_test.go
Comment (arvindbr8) [reply internal/transport/transport_test.go:557]: Oh oh I see. We are relying on close(s.error) to exit the blocking read here. In that case, is there a reason to keep s.ready?

File: internal/transport/keepalive_test.go
Comment (arvindbr8) [inline internal/transport/keepalive_test.go:318]: Using enums for PingMode might allow us to nix this comment.

**Review by arvindbr8 — CHANGES_REQUESTED** [review]
Apologies in the delay. I have gotten back to your changes with a few comments. Please let me know if you want me to elaborate on any comments.

**Comment by tobotg** [comment]
Thanks @arvindbr8 for the review. I'll look into it during this weekend and get back to you if more elaboration needed.

**Pull request closed without merging**
